src/modbus-rtu.c

```diff
@@ -272,32 +272,84 @@ static ssize_t _modbus_rtu_send(modbus_t *ctx, const uint8_t *req, int req_lengt
 #if defined(_WIN32)
     modbus_rtu_t *ctx_rtu = ctx->backend_data;
     DWORD n_bytes = 0;
-    return (WriteFile(ctx_rtu->w_ser.fd, req, req_length, &n_bytes, NULL)) ? (ssize_t)n_bytes : -1;
+    ssize_t rv;
+
+    rv = WriteFile(ctx_rtu->w_ser.fd, req, req_length, &n_bytes, NULL) ? (ssize_t)n_bytes : -1;
+
+    /* if local echo workaround is enabled, try to discard the count of bytes
+     *  we sent ourself */
+    if (ctx_rtu->is_echo_suppressing && n_bytes > 0) {
+        /* we don't use a timeout here since we should have our own caused stuff returned
+         * immediately - and we know the internals of win32_ser_select from code above */
+        if (win32_ser_select(&ctx_rtu->w_ser, (int)n_bytes, NULL) <= 0) {
+             /* propagate errors and timeouts */
+             return -1;
+        } else {
+             /* discard the count of bytes we caused ourself, but keep possible bytes
+              * of following packet we already received */
+             unsigned int n = ctx_rtu->w_ser.n_bytes;
+
+             if (n > (unsigned int)n_bytes) {
+                 n = (unsigned int)n_bytes;
+             }
+
+             ctx_rtu->w_ser.n_bytes -= n;
+        }
+    }
+    return rv;
 #else
-#if HAVE_DECL_TIOCM_RTS
     modbus_rtu_t *ctx_rtu = ctx->backend_data;
-    if (ctx_rtu->rts != MODBUS_RTU_RTS_NONE) {
-        ssize_t size;
+    ssize_t write_size;
 
+#if HAVE_DECL_TIOCM_RTS
+    if (ctx_rtu->rts != MODBUS_RTU_RTS_NONE) {
         if (ctx->debug) {
             fprintf(stderr, "Sending request using RTS signal\n");
         }
 
         ctx_rtu->set_rts(ctx, ctx_rtu->rts == MODBUS_RTU_RTS_UP);
         usleep(ctx_rtu->rts_delay);
+    }
+#endif
 
-        size = write(ctx->s, req, req_length);
+    write_size = write(ctx->s, req, req_length);
 
+#if HAVE_DECL_TIOCM_RTS
+    if (ctx_rtu->rts != MODBUS_RTU_RTS_NONE) {
         usleep(ctx_rtu->onebyte_time * req_length + ctx_rtu->rts_delay);
         ctx_rtu->set_rts(ctx, ctx_rtu->rts != MODBUS_RTU_RTS_UP);
-
-        return size;
-    } else {
-#endif
-        return write(ctx->s, req, req_length);
-#if HAVE_DECL_TIOCM_RTS
     }
 #endif
+
+    /* if local echo workaround is enabled, try to discard the count of bytes
+     * we sent ourself */
+    if (ctx_rtu->is_echo_suppressing && write_size > 0) {
+        uint8_t req_echo[MODBUS_RTU_MAX_ADU_LENGTH];
+        ssize_t read_size = 0;
+
+        while (read_size < write_size) {
+            ssize_t count = read(ctx->s, &req_echo[read_size], write_size - read_size);
+
+            /* return immediately on error */
+            if (count < 0) {
+                return -1;
+            }
+
+            /* 'c' can also be zero - which usually indicates EOF - but this means
+             * in our case here that the echo-ed bytes are not yet available for reading
+             * back. We do busy waiting here to dispatch them as fast as possible. */
+            read_size += count;
```

Comment:
I guess you mean `'count'` instead of `'c'`, right?

Why is it okay to ignore `EOF` here? If the serial port vanishes into thin air (e.g. because of a USB error in case of USB-RS485 dongle) the application using libmodbus will busy wait forever, right?

Reply:
Yepp, seems that I started with `c`, then renamed to `count` to be more meaningful and forgot to fix the comment...

> If the serial port vanishes into thin air (e.g. because of a USB error in case of USB-RS485 dongle) the application using libmodbus will busy wait forever, right?

Shouldn't this scenario result in -1 because the `ctx->s` becomes invalid?

Reply:
If the port is closed `read`'s return code will be 0, the [errors](https://man7.org/linux/man-pages//man2/read.2.html#ERRORS) do not apply in that case. The same can be observed when connecting to the port with [microcom](https://github.com/pengutronix/microcom) and then the port is being closed:

`Got EOF from port`

[Source](https://github.com/pengutronix/microcom/blob/cbeda02ef0b63407f86b0696f91d6be58ba4e070/mux.c#L474)

So again, why should we ignore that and busy wait forever?